controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -535,25 +583,25 @@ func (r *KubeadmControlPlaneReconciler) generateKubeadmConfig(ctx context.Contex
 	return bootstrapRef, nil
 }
 
-func (r *KubeadmControlPlaneReconciler) failureDomainForScaleUp(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, cluster *clusterv1.Cluster) (*string, error) {
+func (r *KubeadmControlPlaneReconciler) failureDomainForScaleDown(cluster *clusterv1.Cluster, machines internal.MachineSet) *string {
```

Comment:
I know this is not in plan, but just want to understand: 
is it possible that cluster.Status.FailureDomains got changed, especially the case that some failure domain got removed? Or we require InfraCluster provider should not support remove failure domains for an existing cluster?

Reply:
Removal of failure domains for running clusters is not currently a feature that we intend to support.

That said with the way that the implementation is currently intended to be defined (in order to support going from no failure domains defined to having failure domains defined), if the infrastructure provider changed the defined failure domains, it would not cause any changes to existing Machines. The newly defined failure domains would take effect for subsequent operations on a KubeadmControlPlane, though (such as scale up, scale down, or upgrade in particular). 

This does make me think that we could potentially have an issue in handling scale down properly for this use case, though. We'll need to tweak the behavior a bit to try and target Machines that either have no failure domains defined or that have no longer existing failure domains before Machines that have existing defined failure domains.

Reply:
I've updated the handling of machines not currently in defined failure domains with the latest changes.